files/class-api-client.php

```diff
@@ -0,0 +1,179 @@
+<?php
+
+namespace Automattic\VIP\Files;
+
+use WP_Error;
+
+require( __DIR__ . '/class-curl-streamer.php' );
+
+function new_api_client() {
+	return new API_Client(
+		constant( 'FILE_SERVICE_ENDPOINT' ),
+		constant( 'FILES_CLIENT_SITE_ID' ),
+		constant( 'FILES_ACCESS_TOKEN' )
+	);
+}
+
+class API_Client {
+	const DEFAULT_REQUEST_TIMEOUT = 10;
+
+	private $api_base;
+	private $files_site_id;
+	private $files_token;
+
+	public function __construct( $api_base, $files_site_id, $files_token ) {
+		$api_base = esc_url_raw( $api_base, [ 'https', 'http' ] );
+		$api_base = untrailingslashit( $api_base );
+		$this->api_base = $api_base;
+
+		$this->files_site_id = $files_site_id;
+		$this->files_token = $files_token;
+	}
+
+	public function get_api_url( $path ) {
+		$path = ltrim( $path, '/\\' );
+		return $this->api_base . '/' . $path;
+	}
+
+	private function call_api( $path, $method, $request_args = [] ) {
+		$request_url = $this->get_api_url( $path );
+
+		$headers = [
+			'X-Client-Site-ID' => $this->files_site_id,
+			'X-Access-Token' => $this->files_token,
+		];
+
+		if ( isset( $request_args['headers'] ) ) {
+			$headers = array_merge( $headers, $request_args['headers'] );
+		}
+
+		$timeout = $request_args['timeout'] ?? self::DEFAULT_REQUEST_TIMEOUT;
+
+		$request_args = [
+			'method' => $method,
+			'headers' => $headers,
+			'timeout' => $timeout,
+		];
+
+		$response = wp_remote_request( $request_url, $request_args );
+
+		return $response;
+	}
+
+	// TODO: implement get_unique_filename()
+
+	public function upload_file( $local_path, $upload_path ) {
+		if ( ! file_exists( $local_path ) ) {
+			/* translators: 1: local file path 2: remote upload path */
+			return new WP_Error( 'upload_file-failed-invalid_path', __( 'Failed to upload file `%1$s` to `%2$s`; the file does not exist.' ), $local_path, $upload_path );
+		}
+
+		$file_size = filesize( $local_path );
+		$file_name = basename( $local_path );
+		[
+			'type' => $file_mime,
+		] = wp_check_filetype( $file_name );
```

Comment:
Minor: What do you think about collapsing this to one line?

```
[ 'type' => $file_mime ] = wp_check_filetype( $file_name );
```

It could just be me getting used to new syntax, but it's a little hard to read at first glance.

Reply:
I originally had it as a single line but changed it because PHPCS was complaining. I think you're right that it does look a bit odd, especially with just one element.